### PR TITLE
[Feature](has_all) support has_all funtion 

### PR DESCRIPTION
**Pull request by ZhangYiXi-dev**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Checks whether one array is a subset of another.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/function_has_all.h [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            execute_has_all_num<ColumnInt32>(*src1_nested_column,*src2_nested_column,nested_null_map_1,
+                                               nested_null_map_2,vec_res,offsets1,offsets2);
+        } else if (check_column<ColumnInt64>(*src1_nested_column)) {
+            execute_has_all_num<ColumnInt64>(*src1_nested_column,*src2_nested_column,nested_null_map_1,
+                                               nested_null_map_2,vec_res,offsets1,offsets2);
+        } else if (check_column<ColumnInt128>(*src1_nested_column)) {
+            execute_has_all_num<ColumnInt128>(*src1_nested_column,*src2_nested_column,nested_null_map_1,
+                                               nested_null_map_2,vec_res,offsets1,offsets2);
+        } 
+        else if (check_column<ColumnFloat32>(*src1_nested_column)) {
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:211]: format code

File: be/src/vec/functions/function_has_all.h [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+class FunctionHasAll: public IFunction {
+public:
+    static constexpr auto name = "has_all";
+    using NullMapType = PaddedPODArray<UInt8>;
+    String get_name() const override { return name; }
+    bool is_variadic() const override { return false; }
+    size_t get_number_of_arguments() const override { return 2; }
+    static FunctionPtr create() { return std::make_shared<FunctionHasAll>(); }
+    //bool is_suitable_for_short_circuit_arguments_execution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:39]: remove unused line

File: be/src/vec/functions/function_has_all.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.cpp:2]: put this function under `be/src/vec/functions/array` directory

File: be/src/vec/functions/simple_function_factory.h [hunk 4]
```diff
@@ -86,6 +86,7 @@ void register_function_hex_variadic(SimpleFunctionFactory& factory);
 
 void register_function_url(SimpleFunctionFactory& factory);
 
+void register_function_has_all(SimpleFunctionFactory& factory);
```
Comment (eldenmoon) [inline be/src/vec/functions/simple_function_factory.h:89]: should be registered in `be/src/vec/functions/array/function_array_register.cpp`

File: gensrc/script/doris_builtins_functions.py [hunk 5]
```diff
@@ -2810,6 +2810,11 @@
     # search functions
     [['multi_search_all_positions'], 'ARRAY_INT', ['STRING', 'ARRAY_STRING'], '', '', '', 'vec', 'ALWAYS_NOT_NULLABLE'],
     [['multi_match_any'], 'TINYINT', ['STRING', 'ARRAY_STRING'], '', '', '', 'vec', 'ALWAYS_NOT_NULLABLE'],
+
+    [['has_all'], 'BOOLEAN', ['ARRAY_SMALLINT', 'ARRAY_SMALLINT'], '', '', '', 'vec', ''],
+    [['has_all'], 'BOOLEAN', ['ARRAY_INT', 'ARRAY_INT'], '', '', '', 'vec', ''],
+    [['has_all'], 'BOOLEAN', ['ARRAY_BIGINT', 'ARRAY_BIGINT'], '', '', '', 'vec', ''],
+    [['has_all'], 'BOOLEAN', ['ARRAY_LARGEINT', 'ARRAY_LARGEINT'], '', '', '', 'vec', ''],
 ]
```
Comment (eldenmoon) [inline gensrc/script/doris_builtins_functions.py:2818]: need more types

File: be/src/vec/functions/function_has_all.h [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                    if(src2_null_map&&(*src2_null_map)[i]&&src1_null_flag){
+                        break;
+                    }
+                    if(src2_null_map&&(*src2_null_map)[i]&&!src1_null_flag){
+                        res=false;
+                        break;
+                    }
+                    if(src1_datas[j]==src2_datas[i]){
+                            break;
+                        }
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:132]: format code

File: be/src/vec/functions/function_has_all.h [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+         ColumnPtr src2_column =
+                block.get_by_position(arguments[1]).column->convert_to_full_column_if_const();
+        const auto& src2_column_array = check_and_get_column<ColumnArray>(*src2_column);
+        const auto& offsets2 = src2_column_array->get_offsets();
+        const auto* src2_nested_column = &src2_column_array->get_data();
+        
+        DataTypePtr src1_column_type = block.get_by_position(arguments[0]).type;
+        auto nested1_type = assert_cast<const DataTypeArray&>(*src1_column_type).get_nested_type();
+        DataTypePtr src2_column_type = block.get_by_position(arguments[1]).type;
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:163]: too many duplicated code, need do some refactor

File: be/src/vec/functions/function_has_all.h [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+            pos_2=offset_2;
+        }
+    }
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) override{
+        ColumnPtr src1_column =
+                block.get_by_position(arguments[0]).column->convert_to_full_column_if_const();
+        const auto& src1_column_array = check_and_get_column<ColumnArray>(*src1_column);
+        const auto& offsets1 = src1_column_array->get_offsets();
+        const auto* src1_nested_column = &src1_column_array->get_data();
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:153]: rename these variables, eg `lhs_nested_column` is better than `src1_nested_column`

File: be/src/vec/functions/function_has_all.h [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+#include "vec/data_types/data_type_number.h"
+#include "vec/functions/simple_function_factory.h"
+#include "vec/data_types/data_type.h"
+
+#include "vec/columns/column_const.h"
+#include "vec/functions/function.h"
+#include "vec/functions/function_helpers.h"
+namespace doris::vectorized {
+
+class FunctionHasAll: public IFunction {
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:31]: add a brief annotation for this function, to quick understand this function

File: be/src/vec/functions/function_has_all.h [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    static FunctionPtr create() { return std::make_shared<FunctionHasAll>(); }
+    //bool is_suitable_for_short_circuit_arguments_execution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+    DataTypePtr get_return_type_impl(const DataTypes & arguments) const override{
+       return std::make_shared<DataTypeUInt8>();
+    }
+    template <typename NestedColumnType>
+    void execute_has_all_string(const IColumn& src1_column,const IColumn& src2_column,const NullMapType* src1_null_map,
+        const NullMapType* src2_null_map,ColumnUInt8::Container& result,const ColumnArray::Offsets64& offsets1,const ColumnArray::Offsets64& offsets2){
+            
+            const ColumnString* src1_data_concrete = reinterpret_cast<const ColumnString*>(&src1_column);
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:47]: better `static_cast` or `check_and_get_column`

File: be/src/vec/functions/function_has_all.h [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+                                               nested_null_map_2,vec_res,offsets1,offsets2);
+        } 
+        else if (check_column<ColumnFloat32>(*src1_nested_column)) {
+            execute_has_all_num<ColumnFloat32>(*src1_nested_column,*src2_nested_column,nested_null_map_1,
+                                               nested_null_map_2,vec_res,offsets1,offsets2);
+        } else if (check_column<ColumnFloat64>(*src1_nested_column)) {
+            execute_has_all_num<ColumnFloat64>(*src1_nested_column,*src2_nested_column,nested_null_map_1,
+                                               nested_null_map_2,vec_res,offsets1,offsets2);
+        } 
+        else if (check_column<ColumnDecimal128>(*src1_nested_column)) {
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:218]: add `date` and `datetime` type

File: be/src/vec/functions/function_has_all.h [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+                //check src1_null_map is or not has true
+                bool src1_null_flag=false;
+                for(size_t n = pos_1; n < offset_1; ++n){
+                    if(src1_null_map && (*src1_null_map)[n]){
+                        src1_null_flag=true;
+                        break;
+                    }
+                }
+                for (size_t i = pos_2; i < offset_2; ++i){
+                    for (size_t j = pos_1; j < offset_1; ++j){
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:65]: need to do refactor , triple nested `for` loops is really hard too understand

Comment (eldenmoon) [reply be/src/vec/functions/function_has_all.h:65]: and find a better variable to rename `i` and `j`

File: be/src/vec/functions/function_has_all.h [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+                        }
+                        if(src1_data_concrete->get_data_at(j)==src2_data_concrete->get_data_at(i)){
+                                break;
+                        }
+                        if(j==offset_1-1){
+                                res=false;
+                                break;
+                        }
+                    }
+                    if(res==false)
```
Comment (eldenmoon) [inline be/src/vec/functions/function_has_all.h:84]: `if (!res)`

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.87 seconds
 load time: 448 seconds
 storage size: 17178683450 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221110163952_clickbench_pr_43761.html

File: be/src/vec/functions/array/function_has_all.h [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+
+            ssize_t left_start = (*left_data.offsets_ptr)[row - 1];
+            ssize_t left_size = (*left_data.offsets_ptr)[row] - left_start;
+            ssize_t right_start = (*right_data.offsets_ptr)[row - 1];
+            ssize_t right_size = (*right_data.offsets_ptr)[row] - right_start;
+            if (right_size == 0) {
+                dst_data[row] = 1;
+                continue;
+            }
+
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_has_all.h:208]: if `right_size` > `left_size` return `false`?

File: be/src/vec/functions/array/function_has_all.h [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+            ExecutorImpl impl;
+            impl.insert_array(left_data.nested_col, left_start, left_size);
+            dst_data[row] = impl.find_all(right_data.nested_col, right_start, right_size);
+        }
+        return Status::OK();
+    }
+
+    Status _execute_nullable(const ColumnArrayExecutionData& data, UInt8* dst_nullmap_data) {
+        for (ssize_t row = 0; row < data.offsets_ptr->size(); ++row) {
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_has_all.h:217]: The function only return OK, just return `void`.

File: be/src/vec/functions/array/function_has_all.h [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+            if (dst_nullmap_data[row]) {
+                continue;
+            }
+
+            if (data.array_nullmap_data && data.array_nullmap_data[row]) {
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_has_all.h:222]: `data.array_nullmap_data` check should out of for loop.

**Comment by eldenmoon** [comment]
please rebase

**Pull request closed without merging**